File: svn_client.py
import subprocess
import re
import base64
from pathlib import Path
from datetime import datetime
from typing import List, Dict, Optional, Tuple
from xml.etree import ElementTree
from config import load_config

class SVNClient:
    def __init__(self, svn_path=None):
        self.svn_path = svn_path or load_config().get('svn_path', 'svn')
        self.default_timeout = 1800  # 30分钟
        self.list_timeout = 3600     # 60分钟用于列表操作
        self.content_timeout = 300    # 5分钟用于文件内容
    
    def _run_command(self, args: List[str], username: str = None, password: str = None, 
                     timeout: int = None) -> Tuple[int, str, str]:
        if timeout is None:
            timeout = self.default_timeout
        
        cmd = [self.svn_path] + args
        cmd.extend(['--non-interactive', '--trust-server-cert-failures=unknown-ca'])
        
        if username:
            cmd.extend(['--username', username])
        if password:
            cmd.extend(['--password', password])
        
        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout,
                encoding='utf-8',
                errors='replace'
            )
            return result.returncode, result.stdout, result.stderr
        except subprocess.TimeoutExpired:
            return -1, '', 'Command timeout'
        except FileNotFoundError:
            return -1, '', f'SVN executable not found: {self.svn_path}'
    
    def test_connection(self, url: str, username: str = None, password: str = None) -> Tuple[bool, str]:
        returncode, stdout, stderr = self._run_command(
            ['info', url, '--xml'],
            username, password,
            timeout=30
        )
        
        if returncode == 0:
            return True, 'Connection successful'
        return False, stderr or 'Connection failed'
    
    def list_directory(self, url: str, username: str = None, password: str = None,
                       recursive: bool = True) -> List[Dict]:
        args = ['list', url, '--xml']
        if recursive:
            args.append('-R')
        
        returncode, stdout, stderr = self._run_command(args, username, password, self.list_timeout)
        
        if returncode != 0:
            raise Exception(stderr or 'Failed to list directory')
        
        return self._parse_list_output(stdout, url)
    
    def _parse_list_output(self, xml_output: str, base_url: str) -> List[Dict]:
        files = []
        try:
            root = ElementTree.fromstring(xml_output)
            base_url = base_url.rstrip('/')
            
            for entry in root.findall('.//entry'):
                path = entry.get('path', '')
                kind = entry.get('kind', 'file')
                
                size_elem = entry.find('size')
                size = int(size_elem.text) if size_elem is not None and size_elem.text else 0
                
                commit = entry.find('commit')
                revision = int(commit.get('revision', 0)) if commit is not None else 0
                
                last_modified = None
                if commit is not None:
                    date_elem = commit.find('date')
                    if date_elem is not None and date_elem.text:
                        last_modified = date_elem.text
                
                filename = Path(path).name or path
                if filename.isEmpty() and not path.isEmpty():
                    filename = path
                
                files.append({
                    'path': path,
                    'filename': filename,
                    'is_dir': kind == 'dir',
                    'size': size,
                    'revision': revision,
                    'last_modified': last_modified,
                    'full_url': f'{base_url}/{path}'
                })
        except ElementTree.ParseError as e:
            pass
        
        return files
    
    def get_file_content(self, url: str, username: str = None, password: str = None) -> Tuple[bytes, str]:
        returncode, stdout, stderr = self._run_command(
            ['cat', url],
            username, password,
            self.content_timeout
        )
        
        if returncode != 0:
            raise Exception(stderr or 'Failed to get file content')
        
        return stdout.encode('utf-8', errors='replace'), 'text'
    
    def get_info(self, url: str, username: str = None, password: str = None) -> Optional[Dict]:
        returncode, stdout, stderr = self._run_command(
            ['info', url, '--xml'],
            username, password
        )
        
        if returncode != 0:
            return None
        
        try:
            root = ElementTree.fromstring(stdout)
            entry = root.find('.//entry')
            if entry is not None:
                return {
                    'url': entry.find('url').text if entry.find('url') is not None else None,
                    'revision': entry.get('revision'),
                    'kind': entry.get('kind'),
                    'root': entry.find('root').text if entry.find('root') is not None else None,
                }
        except ElementTree.ParseError:
            pass
        
        return None

svn_client = SVNClient()

File: desktop/src-tauri/src/svn_client.rs
use std::process::Command;
use std::str;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SvnFile {
    pub path: String,
    pub filename: String,
    pub is_dir: bool,
    pub size: u64,
    pub revision: u64,
    pub last_modified: Option<String>,
}

pub struct SvnClient {
    svn_path: String,
}

impl SvnClient {
    pub fn new(svn_path: Option<String>) -> Self {
        Self {
            svn_path: svn_path.unwrap_or_else(|| "svn".to_string()),
        }
    }

    pub fn test_connection(&self, url: &str, username: Option<&str>, password: Option<&str>) -> Result<bool, String> {
        let mut cmd = Command::new(&self.svn_path);
        cmd.args(["info", url, "--xml"]);
        cmd.args(["--non-interactive", "--trust-server-cert"]);
        
        if let Some(user) = username {
            cmd.args(["--username", user]);
        }
        if let Some(pass) = password {
            cmd.args(["--password", pass]);
        }

        let output = cmd.output()
            .map_err(|e| format!("Failed to execute svn: {}", e))?;

        Ok(output.status.success())
    }

    pub fn list_directory(&self, url: &str, username: Option<&str>, password: Option<&str>, recursive: bool) -> Result<Vec<SvnFile>, String> {
        let mut cmd = Command::new(&self.svn_path);
        cmd.args(["list", url, "--xml"]);
        
        if recursive {
            cmd.arg("-R");
        }
        
        cmd.args(["--non-interactive", "--trust-server-cert"]);
        
        if let Some(user) = username {
            cmd.args(["--username", user]);
        }
        if let Some(pass) = password {
            cmd.args(["--password", pass]);
        }

        let output = cmd.output()
            .map_err(|e| format!("Failed to execute svn: {}", e))?;

        if !output.status.success() {
            let stderr = str::from_utf8(&output.stderr).unwrap_or("");
            return Err(format!("SVN list failed: {}", stderr));
        }

        let stdout = str::from_utf8(&output.stdout).unwrap_or("");
        self.parse_list_output(stdout, url)
    }

    fn parse_list_output(&self, xml: &str, base_url: &str) -> Result<Vec<SvnFile>, String> {
        let mut files = Vec::new();
        
        let mut current_depth = 0;
        let mut path_stack = Vec::new();

        for line in xml.lines() {
            if line.contains("<entry") {
                if let Some(path_start) = line.find("path=\"") {
                    if let Some(path_end) = line[path_start + 6..].find("\"") {
                        let path = &line[path_start + 6..path_start + 6 + path_end];
                        
                        let depth = path.matches('/').count();
                        while path_stack.len() > depth {
                            path_stack.pop();
                        }
                        
                        let full_path = if path_stack.is_empty() {
                            path.to_string()
                        } else {
                            format!("{}/{}", path_stack.join("/"), path)
                        };
                        
                        let filename = if let Some(last_slash) = full_path.rfind('/') {
                            full_path[last_slash + 1..].to_string()
                        } else {
                            full_path.clone()
                        };
                        
                        let is_dir = line.contains("kind=\"dir\"");
                        let size = self.extract_size(line);
                        let revision = self.extract_revision(line);
                        
                        files.push(SvnFile {
                            path: full_path,
                            filename,
                            is_dir,
                            size,
                            revision,
                            last_modified: None,
                        });
                        
                        path_stack.push(path.to_string());
                    }
                }
            }
        }

        Ok(files)
    }

    fn extract_size(&self, line: &str) -> u64 {
        if let Some(size_start) = line.find("size=\"") {
            if let Some(size_end) = line[size_start + 6..].find("\"") {
                return line[size_start + 6..size_start + 6 + size_end]
                    .parse()
                    .unwrap_or(0);
            }
        }
        0
    }

    fn extract_revision(&self, line: &str) -> u64 {
        if let Some(rev_start) = line.find("revision=\"") {
            if let Some(rev_end) = line[rev_start + 10..].find("\"") {
                return line[rev_start + 10..rev_start + 10 + rev_end]
                    .parse()
                    .unwrap_or(0);
            }
        }
        0
    }
}
